### PR TITLE
REALMC-11689: add arrow functions, typed array fixes, more date-time formats, ...

### PR DESCRIPTION
**Pull request by Gabri3l**
* Fixed typed arrays' defineProperty and indexing. Fixes https://github.com/dop251/goja/issues/308.
* Fix Proxy creation panicing on target not callable
* Do not modify sb for variadic calls as it breaks local variables resolution. Instead place a marker on stack and use it to count the number of args. Fixes https://github.com/dop251/goja/issues/311.
* Support patterns in catch clause. See https://github.com/dop251/goja/issues/305
* Fixed the handling of Symbol properties in destructuring assignments. Fixes https://github.com/dop251/goja/issues/312.
* Ensure ToPropertyKey happens earlier when assigning computed keys. Fixes https://github.com/dop251/goja/issues/312.
* Aligned detached buffer semantics with the latest specs. Fixes https://github.com/dop251/goja/issues/315.
* More typed arrays fixes
* Fixed accessor and property key function names. Fixes https://github.com/dop251/goja/issues/314.
* Fixed possible panic when sorting non-standard arrays.
* Added nil-safety checks for values returned by get*() methods
* Fixed panics in parser on some invalid inputs. Fixes https://github.com/dop251/goja/issues/318.
* Arrow function (https://github.com/dop251/goja/pull/319)
* Implemented arrow functions. See https://github.com/dop251/goja/issues/304.
* Define the name property for anonymous functions (including arrow functions)
* Treat "arguments" as a lexical binding to match the latest specification
* Allow arrow functions to contain 'use strict' for simple parameter lists. Fixes https://github.com/dop251/goja/pull/323.
* Fixed argument variable reference resolution in stashless functions
* Treat date-only formats as UTC and date-time as local timezone. Added support for additional datetime formats. Fixes https://github.com/dop251/goja/issues/281, fixes https://github.com/dop251/goja/issues/292.
* Use correct createArgsRest variant when arguments are not in stash. Fixes https://github.com/dop251/goja/issues/327
* Fixed formatting for go 1.17
* Report 'length' as own property for Go slices. Fixes https://github.com/dop251/goja/issues/328.

### CODE REVIEW

File: memory_test.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ func (muc TestNativeMemUsageChecker) NativeMemUsage(value interface{}) (uint64,
 func TestMemCheck(t *testing.T) {
 	// This is the sum of property names allocated at each new (empty) scope
 	var emptyFunctionScopeOverhead uint64 = 8
-	var functionStackOverhead uint64 = 30
+	var functionStackOverhead uint64 = 34
```
Comment (Gabri3l) [inline memory_test.go:28]: This is the only change I had to make

**Review by tsedgwick — APPROVED** [review]
LGTM

File: func.go [hunk 2]
```diff
@@ -15,12 +15,23 @@ type baseFuncObject struct {
...
+
+type funcObject struct {
+	baseJsFuncObject
+}
+
+type arrowFuncObject struct {
+	baseJsFuncObject
+	this      Value
+	newTarget Value
 }
```
Comment (arahmanan) [inline func.go:35]: [q] should we modify/add the `MemUsage` function for these objects?

**Review by arahmanan — APPROVED** [review]
LGTM! Just a question, which could be part of a separate ticket/PR since I noticed that some other existing objects such as `mapIterObject` and `setObject` don't have a `MemUsage` function.

File: func.go (hunk 2 above)
Comment (Gabri3l) [reply func.go:35]: Good call, let me take a look!

Comment (Gabri3l) [reply func.go:35]: Actually I will create a separate ticket so that I can first try to pull in the remaining changes from goja.

**Pull request merged**